types/time.go

```diff
@@ -2432,12 +2432,29 @@ func microSeconds(t *MysqlTime, input string, ctx map[string]int) (string, bool)
 }
 
 func yearNumericFourDigits(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
-	v, succ := parseDigits(input, 4)
+	return yearNumericNDigits(t, input, ctx, 4)
+}
+
+func yearNumericTwoDigits(t *MysqlTime, input string, ctx map[string]int) (string, bool) {
+	return yearNumericNDigits(t, input, ctx, 2)
+}
+
+func yearNumericNDigits(t *MysqlTime, input string, ctx map[string]int, n int) (string, bool) {
+	var l int
```

Comment:
how about:
```go
effectiveCount, effectiveValue := 0, true
for effectiveCount+1 <= n {
	// try to parse the next digit.
	value, succeed := parseDigits(input, effectiveCount+1)
	if !succeed {
		break
	}
	effectiveCount++
	effectiveValue = value
}
```

Thus we don't need to parse it again in https://github.com/pingcap/tidb/pull/8517/files#diff-3d97ef53b300d76ad11d0d73a7b7a7ebR2449

Reply:
If parseDigits return false when pasrse first digit, we should return false....
maybe we should add another `if` after this code to decide if we shoud return according to effectiveCount ?

like 
```
effectiveCount, effectiveValue := 0, true
for effectiveCount+1 <= n {
	// try to parse the next digit.
	value, succeed := parseDigits(input, effectiveCount+1)
	if !succeed {
		break
	}
	effectiveCount++
	effectiveValue = value
}
if (effectiveCount == 0) {
    return input, false
}
```

Reply:
have done:)